File: Array/06.FirstMissingPositive/FirstMissingPositive.py
class Solution(object):
    def firstMissingPositive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums: return 1

        length = len(nums) + 1
        for x in range(1, length):
            if x not in nums:
                return x
        return length

File: Array/06.FirstMissingPositive/FirstMissingPositive.cpp

class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int len = nums.size();
        for (int i = 0; i < len; ++i) {
            int tmp = nums[i];
            while (tmp > 0 && tmp <= len && nums[tmp-1] != tmp) {
                nums[i] = nums[tmp-1];
                nums[tmp-1] = tmp;
                tmp = nums[i];
            }
        }
        for (int i = 0; i < len; ++i) {
            if (nums[i]-1 != i) { return i + 1; }
        } 

        return len + 1;
    }
};